include/exadg/operators/navier_stokes_calculators.cpp

```diff
@@ -0,0 +1,93 @@
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#include <exadg/operators/navier_stokes_calculators.h>
+
+namespace ExaDG
+{
+template<int dim, typename Number>
+ShearRateCalculator<dim, Number>::ShearRateCalculator()
+  : matrix_free(nullptr), dof_index_u(0), dof_index_u_scalar(0), quad_index(0)
+{
+}
+
+template<int dim, typename Number>
+void
+ShearRateCalculator<dim, Number>::initialize(dealii::MatrixFree<dim, Number> const & matrix_free_in,
+                                             unsigned int const                      dof_index_u_in,
+                                             unsigned int const dof_index_u_scalar_in,
+                                             unsigned int const quad_index_in)
+{
+  matrix_free        = &matrix_free_in;
+  dof_index_u        = dof_index_u_in;
+  dof_index_u_scalar = dof_index_u_scalar_in;
+  quad_index         = quad_index_in;
+}
+
+template<int dim, typename Number>
+void
+ShearRateCalculator<dim, Number>::compute_shear_rate(VectorType & dst, VectorType const & src) const
+{
+  dst = 0;
+
+  matrix_free->cell_loop(&This::cell_loop, this, dst, src);
+}
+
+template<int dim, typename Number>
+void
+ShearRateCalculator<dim, Number>::cell_loop(dealii::MatrixFree<dim, Number> const & matrix_free,
+                                            VectorType &                            dst,
+                                            VectorType const &                      src,
+                                            Range const & cell_range) const
+{
+  CellIntegratorVector integrator_vector(matrix_free, dof_index_u, quad_index);
+  CellIntegratorScalar integrator_scalar(matrix_free, dof_index_u_scalar, quad_index);
+
+  for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+  {
+    integrator_vector.reinit(cell);
+    integrator_vector.gather_evaluate(src, dealii::EvaluationFlags::gradients);
+
+    integrator_scalar.reinit(cell);
+
+    for(unsigned int q = 0; q < integrator_scalar.n_q_points; q++)
+    {
+      symmetrictensor sym_grad_u = integrator_vector.get_symmetric_gradient(q);
+
+      // Shear rate definition according to Galdi et al., 2008
+      // ("Hemodynamical Flows: Modeling, Analysis and Simulation").
+      // trace(sym_grad_u^2) = sym_grad_u : sym_grad_u
```

Comment:
I do not fully understand this equation (or where to see the factor of 2 from this equation).

Reply:
One usually finds a definition (also in that cited source): shear_rate := sqrt(2 * trace(gradS(u)^2)), but since it is symmetric, we can equivalently use the contraction instead, which is cheaper to compute. I can also put the full formula?